Add support for activity counter per check

EvanKuhn-Affirm opened the pull request:
This adds an 'activity counter' per check.  If enabled, the check will only run when the counter is > 0.

This change also adds an API endpoint:

```
/api/status-checks/activity-counter?id=<id>&name=<name>&action=<incr|decr|reset>
```

which allows a user / program to query, increment, decrement, or reset the counter for a given check.

EvanKuhn-Affirm commented on cabot/rest_urls.py:
```diff
@@ -8,20 +8,30 @@
 router = routers.DefaultRouter()
 
 
```
I made a bunch of cleanup changes to this file.  Not entirely necessary.  They are:

- Add comments to `create_viewset()`
- Use dashes instead of underscores in API names (eg: `status-checks` instead of `status_checks`).  More web-friendly.  Does anything currently use these endpoints?

EvanKuhn-Affirm commented on cabot/urls.py:
```diff
@@ -4,35 +4,58 @@
+    UserProfileUpdateView,
+    ShiftListView,
+    subscriptions,
+    ActivityCounterView,
```
`ActivityCounterVew`was added.  The rest of the changes to the `import` statements are just cleanup.

EvanKuhn-Affirm commented:
Update: got it tested. Enabling / disabling via activity counter works.

mahendra commented:
Wouldn't this require a salt change too?

EvanKuhn-Affirm commented:
@mahendra AFAIK it doesn't require any salt changes, but we will need to deploy it via salt:

https://confluence.team.affirm.com/display/INFRA/Deploying+Cabot

We'll also need to perform the table migrations before deploying the code.

LMK if I'm missing anything.

elainearbaugh commented on cabot/cabotapp/models.py:
```diff
@@ -334,6 +334,18 @@ class StatusCheck(PolymorphicModel):
     )
+    activity_counter = models.PositiveIntegerField(
+        default=0,
+    )
```
I still like the idea of an ActivityCounter as a separate model so we don't have to have extra data for all the checks that don't require it, but don't have a super strong opinion either way.

EvanKuhn-Affirm replied:
Yeah, I agree, however I think if I put these fields in a different model, they won't show up in the check's configuration page (correct?).  If we can figure a way around that, we could have a separate model. I can do some investigation.

elainearbaugh replied:
Can we keep use_activity_counter in the check and add a separate ActivityCounter model with a foreign key to status check that actually tracks the counter value? then when the endpoint is hit, find or create the corresponding activitycounter for the statuscheck and update it. I think it would be good to avoid having a field that's null for almost all statuschecks, and since the user doesn't need to input anything for the activitycounter value we don't need it in the view.

So
```
StatusCheck
   - use_activity_counter
   - etc

ActivityCounter
    - activity_count
    - status_check (ForeignKey to StatusCheck)
```

EvanKuhn-Affirm replied:
Ok sure, i'll look into it.  Seems like a better design.

elainearbaugh commented on cabot/cabotapp/models.py:
```diff
@@ -334,6 +334,18 @@ class StatusCheck(PolymorphicModel):
+                  'incremented or decremented via API call. When incremented above '
+                  'zero, the check may run.  When decremented to zero, the check will '
+                  'not run. This allows external processes to enable or disable a check '
+                  'as needed. (Note: the check must also be marked \'Active\' to run.)',
```
Great explanation, was kind of worried about people being confused about this new field but I think this definitely gives them enough information about it. 

As we continue adding new fields to status checks I think we should definitely try to figure out some kind of "advanced options" flow to simplify things. I'll ticket it

elainearbaugh commented on cabot/cabotapp/tasks.py:
```diff
@@ -63,6 +63,8 @@ def run_status_check(pk):
     checks = models.StatusCheck.objects.filter(active=True).all()
     for check in checks:
+        if check.use_activity_counter and check.activity_counter <= 0:
+            continue
```
Could be useful to add logging here in case we need to debug issues where a check isn't running when we think it should be

EvanKuhn-Affirm replied:
Good idea.  Is `info` log level ok?

elainearbaugh commented on cabot/metricsapp/forms/grafana_elastic.py:
```diff
@@ -20,6 +20,7 @@ class Meta:
             'time_range',
             'retries',
             'frequency',
+            'use_activity_counter',
```
Please add [here](https://github.com/Affirm/cabot/blob/master/cabot/metricsapp/forms/elastic.py#L27) also (there's probably a way to make these configs simpler, sorry)

EvanKuhn-Affirm replied:
done

elainearbaugh commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +726,52 @@ def get(self, request, *args, **kwargs):
+    '''Return a JSON response with an error message'''
+    return json_response({ 'detail': message }, code)
+
+class ActivityCounterView(View):
```
Would be good to add unit tests for this ([APITestCase](http://www.django-rest-framework.org/api-guide/testing/#api-test-cases) may be helpful. it's also used in some Cabot tests [here](https://github.com/Affirm/cabot/blob/master/cabot/cabotapp/tests/tests_basic.py#L35))

EvanKuhn-Affirm replied:
Yeah, I forgot about tests completely.  Will add them!

EvanKuhn-Affirm commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +726,52 @@ def get(self, request, *args, **kwargs):
+
+def json_response(data, code):
+    '''Return a JSON response containing some data'''
+    return HttpResponse(json.dumps(data), status=code, content_type="application/json")
```
fixed

EvanKuhn-Affirm commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +726,52 @@ def get(self, request, *args, **kwargs):
+        if action == 'incr':
+            check.activity_counter += 1
+            check.save()
+            message = 'counter incremented'
```
Yes, good idea

EvanKuhn-Affirm commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +726,52 @@ def get(self, request, *args, **kwargs):
+        elif action == 'decr':
+            if check.activity_counter > 0:
+                check.activity_counter -= 1
+                check.save()
```
Yeah I wasn't sure. I thought that returning errors in that case might cause unnecessary headaches, but I could be convinced either way.

EvanKuhn-Affirm commented on cabot/cabotapp/tests/utils.py:
```diff
@@ -0,0 +1,187 @@
+# -*- coding: utf-8 -*-
```
Contents of this file copied from tests_basic.py.  No changes.

EvanKuhn-Affirm commented on tox.ini:
```diff
@@ -58,6 +58,27 @@ commands =
     bash -euc "docker-compose run --rm web python manage.py test cabot || (docker-compose down && exit 1)"
     bash -euc "docker-compose down"
 
+# For running PostgreSQL-based tests multiple times:
```
Added these two tox targets for (a) setting up psql tests, and (b) running them. Makes testing faster when running multiple times.

EvanKuhn-Affirm commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +736,50 @@ def get(self, request, *args, **kwargs):
     return dictionary.get(key)
+
+
+class ActivityCounterView(View):
```
@elainearbaugh I refactored this a bit to extract action-handling logic into `_handle_action()`.  I didn't use an enum as I didn't feel it added anything useful.  If people feel strongly tho, I'll add it.

elainearbaugh replied:
kk sg

EvanKuhn-Affirm commented:
@elainearbaugh @mahendra I've added tests for activity counters. In doing so I refactored some of the cabotapp tests by breaking up tests_basic.py into multiple files.

Anyway, we now have (a) working activity-counter code, (b) tests passing, (c) flake8 passing.  I think this is ready, unless we want to investigate putting the activity-counter fields in a different model.  I'm not sure how to expose those fields in the view, though, as that's handled automatically (via the `ModelViewSet` mixin class I think).

elainearbaugh commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +736,50 @@ def get(self, request, *args, **kwargs):
+
+        # Handle actions (incr, decr, reset)
+        action = request.GET.get('action', None)
+        message = self._handle_action(check, action)
```
handle the case where the action isn't incr, decr, or reset? with another 404?

elainearbaugh commented on cabot/cabotapp/models.py:
```diff
@@ -381,6 +393,21 @@ def last_result(self):
         except:
             return None
 
+    def should_run(self):
```
👍

EvanKuhn-Affirm commented:
Thanks @elainearbaugh.  I'll update the diff by breaking out the ActivityCounter model.  Also, yes, sorry for keeping the test refactor in this PR; I knew I should have broken it out.  Will do so in the future.

EvanKuhn-Affirm commented:
FYI, I changed this PR to depend on https://github.com/Affirm/cabot/pull/114 (the test code refactor).

EvanKuhn-Affirm commented:
@mahendra @elainearbaugh  I've finished updating this to:

- Create a new `ActivityCounter` model/table.  
- Add an index on the StatusCheck name column.

Let me know how it looks.

Also, do you think we should also handle enforcing uniqueness in this change?  Thanks.

EvanKuhn-Affirm commented on cabot/cabotapp/views.py:
```diff
@@ -726,3 +748,92 @@ def get(self, request, *args, **kwargs):
+
+        if action == 'incr':
+            check.activity_counter.count += 1
+            check.activity_counter.save()
```
Good thought, I've updated to wrap the `get()` method with `@transaction.atomic`
